modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,176 @@
+package fee
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	
+	"github.com/cosmos/ibc-go/modules/apps/29-fee/keeper"
+	"github.com/cosmos/ibc-go/modules/apps/29-fee/types"
+	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/modules/core/05-port/types"
+	ibcexported "github.com/cosmos/ibc-go/modules/core/exported"
+)
+
+// IBCModule implements the ICS26 callbacks for the fee middleware given the fee keeper and the underlying application.
+type IBCModule struct {
+	keeper keeper.Keeper
+	app    porttypes.IBCModule
+}
+
+// NewIBCModule creates a new IBCModule given the keeper and underlying application
+func NewIBCModule(k keeper.Keeper, app porttypes.IBCModule) IBCModule {
+	return IBCModule{
+		keeper: k,
+		app:    app,
+	}
+}
+
+// OnChanOpenInit implements the IBCModule interface
+func (im IBCModule) OnChanOpenInit(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version string,
+) error {
+	mwVersion, appVersion := channeltypes.SplitChannelVersion(version)
+	// Since it is valid for fee version to not be specified, the above middleware version may be for a middleware
+	// lower down in the stack. Thus, if it is not a fee version we pass the entire version string onto the underlying
+	// application.
+	// If an invalid fee version was passed, we expect the underlying application to fail on its version negotiation.
+	if mwVersion == types.Version {
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
```

Comment:
Note since we no longer require that the fee version is specified, it's possible that the middleware version we get is actually for some middleware lower in the stack. So we must pass the full version on in this case.

Note: For a stack `fee -> whitelist -> transfer`

the fee middleware cannot differentiate between cases:

1. `wrongfee29-A:whitelist:ics20-1`
2. `whitelist:ics20-1`

So in both cases, it will assume fee is not enabled and pass the full version to the underlying app.

However in case 1, the underlying app will fail its version negotiation since the wrongfee string is unexpected.